pkg/router/template/limiter/limiter.go

```diff
@@ -0,0 +1,144 @@
+	changeReqTime *time.Time
+
+	// handlerRunning indicates whether the Handler is actively running.
+	handlerRunning bool
+
+	// lock protects the CoalescingSerializingRateLimiter structure from multiple threads manipulating it at once.
+	lock sync.Mutex
+
+	// callbackTimer is the timer we use to make callbacks to re-run the function to decide if we need to do work.
+	callbackTimer *time.Timer
+}
+
+func NewCoalescingSerializingRateLimiter(interval time.Duration, handlerFunc HandlerFunc) *CoalescingSerializingRateLimiter {
+	limiter := &CoalescingSerializingRateLimiter{
+		handlerFunc:    handlerFunc,
+		callInterval:   interval,
+		lastStart:      time.Time{},
+		changeReqTime:  nil,
+		handlerRunning: false,
+	}
+
+	return limiter
+}
+
+// RegisterChange() indicates that the rate limited function should be called. It may not immediately run it, but it will cause it to run within
+// the ReloadInterval.  It will always immediately return, the function will be run in the background.  Not every call to RegisterChange() will
+// result in the function getting called.  If it is called repeatedly while it is still within the ReloadInterval since the last run, it will
+// only run once when the time allows it.
+func (csrl *CoalescingSerializingRateLimiter) RegisterChange() {
+	glog.V(8).Infof("RegisterChange called")
+
+	csrl.changeWorker(true)
+}
+
+func (csrl *CoalescingSerializingRateLimiter) changeWorker(userChanged bool) {
+	csrl.lock.Lock()
+	defer csrl.lock.Unlock()
+
+	glog.V(8).Infof("changeWorker called")
+
+	if userChanged && csrl.changeReqTime == nil {
+		// They just registered a change manually (and we aren't in the middle of a change)
+		now := time.Now()
+		csrl.changeReqTime = &now
+	}
+
+	if csrl.handlerRunning {
+		// We don't need to do anything else... there's a run in progress, and when it is done it will re-call this function at which point the work will then happen
+		glog.V(8).Infof("The handler was already running (%v) started at %s, returning from the worker", csrl.handlerRunning, csrl.lastStart.String())
+		return
+	}
+
+	if csrl.changeReqTime == nil {
+		// There's no work queued so we have nothing to do.  We should only get here when
+		// the function is re-called after a reload
+		glog.V(8).Infof("No invoke requested time, so there's no queued work.  Nothing to do.")
+		return
+	}
+
+	// There is no handler running, let's see if we should run yet, or schedule a callback
+	now := time.Now()
+	sinceLastRun := now.Sub(csrl.lastStart)
+	untilNextCallback := csrl.callInterval - sinceLastRun
+	glog.V(8).Infof("Checking reload; now: %v, lastStart: %v, sinceLast %v, limit %v, remaining %v", now, csrl.lastStart, sinceLastRun, csrl.callInterval, untilNextCallback)
+
+	if untilNextCallback > 0 {
+		// We want to reload... but can't yet because some window is not satisfied
+		if csrl.callbackTimer == nil {
+			csrl.callbackTimer = time.AfterFunc(untilNextCallback, func() { csrl.changeWorker(false) })
```

Comment:
At some point we should have a for loop instead of a recursive call, just to avoid the remote possibility of constant changes causing stackoverflow.